chainer/functions/normalization/group_normalization.py

```diff
@@ -1,10 +1,348 @@
+            self.inv_std = xp.empty_like(dummy_beta)
+        x_hat = cudnn.batch_normalization_forward_training(
+            x, self.dummy_gamma, dummy_beta, dummy_beta, dummy_beta,
+            self.mean, self.inv_std, self.eps, 1.0,
+            True, libcudnn.CUDNN_BATCHNORM_SPATIAL,
+            configuration.config.debug)
+
+        y = x_hat.reshape((batch_size, channels, -1))
+        cuda.elementwise(
+            'T gamma, T beta', 'T y',
+            'y = y * gamma + beta',
+            'groupnorm_y')(gamma[:, None], beta[:, None], y)
+
+        y = y.reshape(orig_shape)
+        return y,
+
+    def backward(self, indexes, grad_outputs):
+        x, gamma = self.get_retained_inputs()
+        gy, = grad_outputs
+
+        orig_shape = x.shape
+        batch_size = orig_shape[0]
+        groups = self.groups
+        reduced_shape = (batch_size * groups, -1)
+        x = chainer.functions.reshape(x, reduced_shape)
+
+        x_hat, = _XHat(
+            self.eps, self.mean, self.inv_std,
+            self.dummy_gamma).apply((x,))
+        gx_hat, ggamma, gbeta = _GradHelper().apply((x_hat, gamma, gy))
+        gx, = _XHatGrad(
+            self.eps, self.mean, self.inv_std,
+            self.dummy_gamma, x_hat.array).apply((x, gx_hat))
+
+        gx = gx.reshape(orig_shape)
+        return gx, ggamma, gbeta
+
+
+class _GradHelper(function_node.FunctionNode):
+
+    def forward(self, inputs):
+        self.retain_inputs((0, 1, 2))
+        x_hat, gamma, gy = inputs
+        xp = backend.get_array_module(x_hat)
+
+        reduced_shape = x_hat.shape
+        batch_size, channels = gy.shape[:2]
+        x_hat = x_hat.reshape((batch_size, channels, -1))
+        gy = gy.reshape((batch_size, channels, -1))
+
+        gx_hat = gy * gamma[:, None]
+        if xp is numpy:
+            ggamma = (gy * x_hat).sum(axis=(0, 2))
+        else:
+            ggamma = cuda.reduce(
+                'T gy, T x_hat', 'T ggamma',
+                'gy * x_hat', 'a + b', 'ggamma = a', '0',
+                'groupnorm_ggamma')(gy, x_hat, axis=(0, 2))
+        gbeta = gy.sum(axis=(0, 2))
+
+        gx_hat = gx_hat.reshape(reduced_shape)
+        return gx_hat, ggamma, gbeta
+
+    def backward(self, indexes, grad_outputs):
+        x_hat, gamma, gy = self.get_retained_inputs()
+        ggx_hat, gggamma, ggbeta = grad_outputs
+
+        reduced_shape = x_hat.shape
+        orig_shape = gy.shape
+        batch_size, channels = gy.shape[:2]
+        x_hat = x_hat.reshape((batch_size, channels, -1))
+        gy = gy.reshape((batch_size, channels, -1))
+        ggx_hat = ggx_hat.reshape((batch_size, channels, -1))
+
+        gx_hat2 = gggamma[:, None] * gy
+        ggamma2 = chainer.functions.sum(ggx_hat * gy, axis=(0, 2))
+        ggy = (ggx_hat * gamma[:, None] + gggamma[:, None] * x_hat +
+               ggbeta[:, None])
+
+        gx_hat2 = chainer.functions.reshape(gx_hat2, reduced_shape)
+        ggy = chainer.functions.reshape(ggy, orig_shape)
```

Comment:
Just curious but how come you're using the free function `reshape` instead of the method here (you're mostly using the method in other places)?